llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -231,6 +231,48 @@ static KernelMapEntryScope selectDeviceCodeSplitScopeAutomatically(Module &M) {
   return Scope_PerModule;
 }
 
+// Collect all the dependencies for the function.
+static void collectFunctionCallGraphNodes(llvm::Function *Func,
+                                          SetVector<const GlobalValue *> &GVs) {
+  if (!GVs.count(Func))
+    GVs.insert(Func);
+
+  std::vector<llvm::Function *> Workqueue;
+  Workqueue.push_back(Func);
+
+  while (!Workqueue.empty()) {
+    Function *F = &*Workqueue.back();
+    Workqueue.pop_back();
+    for (auto &I : instructions(F)) {
+      if (CallBase *CB = dyn_cast<CallBase>(&I))
+        if (Function *CF = CB->getCalledFunction())
+          if (!CF->isDeclaration() && !GVs.count(CF)) {
+            GVs.insert(CF);
+            Workqueue.push_back(CF);
+          }
+    }
+  }
+}
+
+// This function returns the collection of functions that are not
+// called from any other function in the module. Example of such
+// an unreferenced function can be a SYCL_EXTERNAL function that
+// will be referenced later, e.g. during linking SPIRV modules.
+static std::vector<llvm::Function *> collectUnreferencedFuncs(Module &M) {
+  SetVector<const GlobalValue *> ReferencedFuncs;
+  for (auto &F : M.functions()) {
+    if (F.getCallingConv() == CallingConv::SPIR_KERNEL)
+      collectFunctionCallGraphNodes(&F, ReferencedFuncs);
+  }
+
+  std::vector<llvm::Function *> UnreferencedFuncs;
+  for (auto &F : M.functions()) {
+    if (!F.isDeclaration() && !ReferencedFuncs.count(&F))
```

Comment:
What if we have an `internal` (I mean linkage type) unreferenced function here? Could it cause any conflicts later?

Reply:
I think we can ignore unreferenced `internal` functions, just like we did before. AFAIK if two LLVM modules defined internal functions with same names, they won't cause conflicts when these two LLVM modules linked. I'm not sure about SPIR-V though, I can't find any analog of `internal` linkage type in SPIR-V spec, but the translator seems to support some Internal linkage type. (@AlexeySotkin , any idea why?).

Reply:
There might be cases when we want to preserve unreferenced functions in SPIR-V, e.g. those which are mentioned in `@llvm.used` variable.
Whether they will conflict or not depends on linkage type.
`internal` linkage type is the default in SPIR-V, i.e. when global variable or function doesn't have `LinkageType`  decoration it has internal linkage.